src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -30,70 +30,99 @@ public interface IPeerSelector
 
         /// <summary>
         /// Constructor for the peer selector.
         /// </summary>
-        /// <param name="peerAddressManager">The singleton peer address manager instance.</param>
-        public PeerSelector(IPeerAddressManager peerAddressManager)
+        /// <param name="peerAddresses">The collection of peer address as managed by the peer address manager.</param>
+        public PeerSelector(ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses)
         {
-            Guard.NotNull(peerAddressManager, nameof(peerAddressManager));
+            Guard.NotNull(peerAddresses, nameof(peerAddresses));
 
-            this.peerAddressManager = peerAddressManager;
+            this.peerAddresses = peerAddresses;
         }
 
         /// <inheritdoc/>
         public PeerAddress SelectPeer()
         {
-            if (this.peerAddressManager.Peers.Tried().Any() == true &&
-                (this.peerAddressManager.Peers.New().Any() == false || new Random().Next(2) == 0))
-                return this.peerAddressManager.Peers.Tried().Random();
+            var tried = this.peerAddresses.Attempted().Concat(this.peerAddresses.Connected());
+            if (tried.Any() == true && (!this.peerAddresses.Fresh().Any() || new Random().Next(2) == 0))
+                return tried.Random();
 
-            if (this.peerAddressManager.Peers.New().Any() == true)
-                return this.peerAddressManager.Peers.New().Random();
+            if (this.peerAddresses.Fresh().Any())
+                return this.peerAddresses.Fresh().Random();
 
             return null;
         }
 
         /// <inheritdoc/>
         public IEnumerable<PeerAddress> SelectPeers()
         {
-            return this.peerAddressManager.Peers.Where(p => p.Value.Preferred).Select(p => p.Value);
+            return this.peerAddresses.Where(p => p.Value.Preferred).Select(p => p.Value);
         }
     }
 
-    /// <summary>
-    /// Extension methods that helps group a set of peers by various attempted,
-    /// connected or handshaked states.
-    /// </summary>
     public static class PeerSelectorExtensions
     {
         /// <summary>
-        /// Return peers where they have never had a connection attempt or have been connected to.
+        /// Return peers which've had connection attempts but none successful. 
+        /// <para>
+        /// The result filters out peers which satisfies the above condition within the 
+        /// last 60 seconds and that has had more than 10 failed attempts.
+        /// </para>
         /// </summary>
-        public static IEnumerable<PeerAddress> New(this ConcurrentDictionary<IPEndPoint, PeerAddress> peers)
+        public static IEnumerable<PeerAddress> Attempted(this ConcurrentDictionary<IPEndPoint, PeerAddress> peers)
         {
-            var isNew = peers.Skip(0).Where(p => p.Value.IsNew).Select(p => p.Value);
-            return isNew;
+            var result = peers.Skip(0).Where(p =>
```

Comment:
I think we should also order the list so that more recent attempts are not in the top.

Reply:
@dangershony sorry, this will also be dropped with next release...